test/unit/storage/LockingResourceStore.test.ts

```diff
@@ -0,0 +1,70 @@
+import { Lock } from '../../../src/storage/Lock';
+import { LockingResourceStore } from '../../../src/storage/LockingResourceStore';
+import { ResourceLocker } from '../../../src/storage/ResourceLocker';
+import { ResourceStore } from '../../../src/storage/ResourceStore';
+
+describe('A LockingResourceStore', (): void => {
+  let store: LockingResourceStore;
+  let locker: ResourceLocker;
+  let lock: Lock;
+  let release: () => Promise<void>;
+  let source: ResourceStore;
+
+  beforeEach(async(): Promise<void> => {
+    source = {
+      getRepresentation: jest.fn(async(): Promise<any> => null),
+      addResource: jest.fn(async(): Promise<any> => null),
+      setRepresentation: jest.fn(async(): Promise<any> => null),
+      deleteResource: jest.fn(async(): Promise<any> => null),
+      modifyResource: jest.fn(async(): Promise<any> => null),
+    };
+    release = jest.fn();
+    locker = {
+      acquire: jest.fn(async(): Promise<any> => {
+        lock = { release };
+        return lock;
+      }),
+    };
+    store = new LockingResourceStore(source, locker);
+  });
+
+  it('acquires a lock on the resource when getting it.', async(): Promise<void> => {
+    await store.getRepresentation({ path: 'path' }, null);
```

Comment:
We should also test the asynchronous behavior; i.e., `release` should only be called _after_ `await`, and I don't think it is at the moment (as per my comment above).

Reply:
Yea, this is something I was also thinking about. But I couldn't actually figure out how to do this cleanly 😅

Reply:
Ah well, `delayedResolve` does it indeed, but what I usually do is storing the `resolve` functions, and then calling them in a controlled way. Something like:

```
const promise = store.getRepresentation({ path: 'path' }, null);
expect(locker.acquire).toHaveBeenCalledTimes(1);
expect(locker.acquire).toHaveBeenLastCalledWith({ path: 'path' });
expect(source.getRepresentation).toHaveBeenCalledTimes(1);
expect(lock.release).toHaveBeenCalledTimes(0); // <== not released yet!
await tellSourceToResolveRepresentationPromise(); // <== assuming `delayedResolve` stores the promises you need rather than calling them itself
expect(lock.release).toHaveBeenCalledTimes(1); // <== only now it is released!
expect(await promise).toEqual(whateverYouReturned); // <== currently not checked
```

Such active control makes it a bit easier.

But the explicit order approach also works.